pkg/gadgets/interface.go

```diff
@@ -16,18 +16,45 @@ package gadgets
 
 import (
 	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+	apimachineryruntime "k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 type Trace interface {
-	Operation(trace *gadgetv1alpha1.Trace, resolver Resolver, operation string, params map[string]string)
+	Operation(trace *gadgetv1alpha1.Trace,
+		resolver Resolver,
+		client client.Client,
+		operation string,
+		params map[string]string)
```

Comment:
Does it make sense to pass `client.Client` to the `TraceFactory` instead of the `Trace`? (I'm just wondering, I don't have a strong opinion).

Reply:
I think it is better to pass the trace, so the controller allows the gadget to only modify the trace resource by patching it. In pseudo-code:
```
	traceBeforeOperation := trace.DeepCopy()
	patch := client.MergeFrom(traceBeforeOperation)
	Operation(trace)
	err = r.Client.Status().Patch(ctx, trace, patch)
```

Reply:
I think I don't follow your reply. What I mean is that the client for every single operation should be the same, so why not to pass it to the factory, so the factory can set it as a member on all Trace instances it creates?

Reply:
Oh, I misunderstood you. Yes, we could do that and probably for the `resolver` argument too.

Reply:
Done. Note that the client is passed to the factories after the controller is started because the client comes from the controller. But the factories are created first because we need to call `factoryWithScheme.AddToScheme(scheme)` before starting the controller.